### ISSUE TITLE
urp pink problem and sprite animation bug in preview window (Two different bugs)

### ISSUE THREAD

**Issue by meredoth** [KybernetikGames/animancer#53]
# Environment

- Animancer Version Number: 4.4.0
- Animancer Pro or Lite: Animancer Pro
- Unity Version: 2020.1.0b10
- Platform: Windows

# Description

1) preview in animancer window is pink 
2) 2d sprite is not shown properly in preview window if the sprite renderer component is in a child object of the animancer component object. 

# Reproduction

1) The urp pink window problem exists with unity 2020.1.0b10 and urp version 8.1.0, I managed to solve it by creating and loading a new unlit shader in the floor class constructor in TransitionPreviewWindow.cs file.

2) The sprite renderer bug happens when the sprite renderer component is in a child gameobject of the animancer component object. I changed the GetAnimationType method in AnimationType.cs to check in the return statement if animator.GetComponentInChildren<SpriteRenderer>() != null but maybe this introduces a new bug in 3d projects. Maybe there should be a field in the inspector for the user to check if the animationtype is sprite ?

**Comment by KybernetikGames** [comment]
Someone else reported the URP issue a while ago and told me they fixed it by adding this to the Floor constructor:
```
var shader = EditorGUIUtility.Load("Previews/PreviewPlaneWithShadow.shader") as Shader;
if (shader == null)
{
	shader = EditorGUIUtility.Load("Previews/Preview_LightColorNode.shader") as Shader;// URP Fallback.
	if (shader == null)
		return;
}
```
I haven't had a chance to verify it though. Is that close to what you did?

The Sprite issue sounds pretty straightforward. Does your Hierarchy look like this?

- Parent: Animator and AnimancerComponent
  - Child: SpriteRenderer

**Comment by meredoth** [comment]
for the floor constructor i changed the var shader to var shader = EditorGUIUtility.Load("Assets/UnlitShader.shader") as Shader; where unlitshader is a shader i created and saved in assets from Assets->Create->shader->unlit shader

the above solution doesn't work for me but replacing the Previews/PreviewPlaneWithShadow.shader with Previews/Preview_LightColorNode.shader works. It seems shader != null at the first check so the if statement body never executes to load the Previews/Preview_LightColorNode.shader

for the sprite issue yes, i usually have the renderer in child object, i could also put the animancer component there or both at parent object or use the solution i described above, just wrote about this since i didn't read anywhere in the manual that the renderer and the animancer component should be on the same gameobject and didn't know if that was the intended usage.

Thanks for the quick reply.

**Comment by KybernetikGames** [comment]
Unity 2019.3 (URP 7.1.8) has no problem with the floor shader and I don't have 2020.1 installed. Would you mind checking if `shader.isSupported` returns false for the broken shader? It says it's for hardware compatibility, but I'm hoping it will also include render pipeline compatibility. If not, I'm thinking I'll have to just give the window a field where you can assign a custom floor material if necessary.

**Comment by meredoth** [comment]
Previews/PreviewPlaneWithShadow.shader
Unity:2020.1  urp: 8.1.0, my project sprite animation : shader.isSupported returns **true** , shows **pink** screen  in animation preview
Unity 2020.1 old renderer(no srp) examples basic 3 sequence coroutine: shader.isSupported returns true, works fine  in animation preview

Previews/Preview_LightColorNode.shader
Unity:2020.1  urp: 8.1.0  my project sprite animation: shader.isSupported throws null reference exception , works fine in animation preview because the if(shader == null) check exits the floor constructor and the floor is never created.
Unity 2020.1 old renderer(no srp) examples basic 3 sequence coroutine: shader.isSupported throws null reference exception works almost fine  in animation preview (shows full grey, instead of grey with white grid) because the if(shader == null) check exits the floor constructor and the floor is never created.

So the Previews/Preview_LightColorNode.shader doesn't work in any case i tried (returns null) but 
the animation preview shows without floor (seems ok in 2d, missing floor in 3d). 
shader.isSupported cannot be used because it returns true even if it shows pink screen with the Previews/PreviewPlaneWithShadow.shader in urp 8.1.0.

A solution might be something like:
```c#
Shader shader = null;
if (!GraphicsSettings.currentRenderPipeline) // if we are in old classic renderer
{
    shader = EditorGUIUtility.Load("Previews/PreviewPlaneWithShadow.shader") as Shader;
}
else
{
    //if we are in any scriptable render pipeline apply shader with field here code
}
if (shader == null)  // else don't render floor
    return;
```
since there are many scriptable render pipelines (urp, hdrp, custom), with different renderers(ex. 2d renderer) and are going to change often in the future.

**Comment by KybernetikGames** [comment]
Thanks for checking that.

I'll probably just have it use the custom floor if one is set, regardless of pipeline. An option to just disable it if you don't care is probably a good idea too. I just wish there was a way to detect that a specific shader won't work so I could explain what to do instead of inevitably getting bug reports about it.

**Comment by meredoth** [comment]
I found out that 
```c#
Shader shader = null;
if (!GraphicsSettings.currentRenderPipeline) // if we are in old classic renderer
{
          shader = EditorGUIUtility.Load("Previews/PreviewPlaneWithShadow.shader") as Shader;
}
else
{
          shader = GraphicsSettings.renderPipelineAsset.terrainDetailGrassBillboardShader;
           //if we are in any scriptable render pipeline 
 }
if (shader == null) // else don't render floor
          return;
 ```
works with the urp, but had to set the const float FloorScale to 1 because the tiles were too large in urp and in sprite mode, the side effect is that the tiles are small in the old renderer but i don't think that matters a lot.
The problem is that this still doesn't work in HDRP, the shader is null so the floor isn't rendered there.

Tested with unity 2020.1.0b10, windows platform, urp 8.1.0, hdrp 8.1.0
urp animationtype sprite (my project) and
old renderer, urp and hdrp with the basics 3 example sequence coroutine.

I think a field where we can put our own shaders if there is no floor would be nice and if it is pink atleast we would know it is our shader problem in that pipeline and change it, since the line 
shader = GraphicsSettings.renderPipelineAsset.terrainDetailGrassBillboardShader; doesn't throw an exception at scriptable render pipelines, it just makes the shader variable null if it can't find the shader and with the if(shader==null) statement exits the constructor without drawing the floor. (In my case the HDRP test).
